src/patch-builtins.js

```diff
@@ -137,9 +137,25 @@ let fragmentMixin = {
    * @this {DocumentFragment}
    */
   querySelectorAll(selector) {
-    return mutation.query(this, function(n) {
-      return utils.matchesSelector(n, selector);
-    });
+    var that =  this
+    if (this instanceof Document) {
+      flush();
+      let result = nativeDQSA.call(this, selector);
+      result = Array.prototype.filter.call(result, function (n) {
+        return !utils.ownerShadyRootForNode(n);
+      });
+      return result;
+    } else if (!this.__shady && this instanceof Element) {
```

Comment:
This codepath is puzzling to me. Could you add clarifying comments to explain the condition and why the filter is applied to childNodes later?

E.g. why not loop through the childNodes and check for matchesselector? That would prevent looping through all other deep children.